src/broadcaster.rs

```diff
@@ -0,0 +1,192 @@
+        if let Async::Ready(value) = shared_state.poll_change(cx) {
+            // value was just broadcast
+            self.state.status.lock().unwrap().has_changed = false;
+            Async::Ready(value)
+        } else {
+            let mut status = self.state.status.lock().unwrap();
+            if status.has_changed {
+                // a new value was polled and broadcast by another task since we last polled
+                status.has_changed = false;
+                Async::Ready(shared_state.value.clone())
+            } else {
+                // nothing new to report
+                status.waker = Some(cx.waker().clone());
+                Async::Pending
+            }
+        }
+    }
+}
+
+impl<A: Signal> Clone for Broadcaster<A>
+    where <A as Signal>::Item: Clone {
+    fn clone(&self) -> Self {
+
+        let new_status = Arc::new(Mutex::new(BroadcasterStatus {
+            has_changed: self.state.status.lock().unwrap().has_changed,
+            waker: None
+        }));
+
+        let shared_state = self.state.shared_state.lock().unwrap();
+        shared_state.notifier.targets.lock().unwrap().push(Arc::downgrade(&new_status));
+
+        let new_state = Arc::new(BroadcasterState {
+            status: new_status,
+            shared_state: self.state.shared_state.clone()
+        });
+
+        Self {
+            state: new_state,
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
+
+// State for a single broadcaster instance.
+//
+// It's split in this way because the BroadcasterSharedState also needs to access the status
+// (e.g. to notify of changes).
+struct BroadcasterState<A: Signal> {
+    status: Arc<Mutex<BroadcasterStatus>>,
+    shared_state: Arc<Mutex<BroadcasterSharedState<A>>>,
+}
+
+// ---------------------------------------------------------------------------------------------
+
+struct BroadcasterStatus {
+    has_changed: bool,
+    waker: Option<Waker>
+}
+
+// ---------------------------------------------------------------------------------------------
+
+// Shared state underpinning a Cloned set of broadcasters
+struct BroadcasterSharedState<A: Signal> {
+    signal: A,
+    value: Option<A::Item>,
+    notifier: Arc<BroadcasterNotifier>
+}
+
+impl<A: Signal> BroadcasterSharedState<A>
+        where <A as Signal>::Item: Clone {
+
+    // Poll the underlying signal for changes, giving it a BroadcasterWaker to wake in
+    // the future if it is in Pending state.
+    //
+    // Takes self by Arc<Mutex<Self>> because this construct is needed to build a
+    // BroadcasterWaker
+    fn poll_change(&mut self, cx: &mut Context) -> Async<Option<A::Item>> {
+
+        // Set the notifier to be waiting on a poll, and if it wasn't previously, actually
+        // execute that poll.
+        //
+        // TODO: Do we actually need AcqRel guarantee?
+        if !self.notifier.is_waiting.swap(true, Ordering::AcqRel) {
```

Comment:
As a general rule, I prefer using `SeqCst`. After *thoroughly* testing that it works correctly in a multi-threaded environment (including lots of unit tests), then it can be relaxed to a weaker mode.

Reply:
(That also goes for all the other uses of `Atomic*`)